### DISCUSSION TITLE
Why are upstream manifests vendored?

### DISCUSSION THREAD

**Discussion by jaklan** [deployKF/deployKF#15] [Questions & Answers] [answered]
Hello,

I have a quick question - I have noticed all `kubeflow-tools` manifests are stored directly in the repo in the `upstream` directories and synced with `sync_upstream.sh` script. What's the goal of that approach instead of referring to the original manifests directly (so you could i.a. have more flexibility on upstream refs you want to use)?

Regards!

**Comment by thesuperzapper** [comment, marked as answer]
@jaklan thanks for the question!

deployKF initially started with the approach of [directly referencing the base Kustomize `resources`](https://kubectl.docs.kubernetes.io/references/kustomize/kustomization/resource/) from git URLs.
However, there were some significant issues with this approach:

1. This prevented the generated manifests from being used if GitHub is not accessible (for example in air-gapped environments):
     - _NOTE: for the "vendored" helm charts, you can easily use a private helm repo by changing values like [`deploykf_dependencies.cert_manager.charts.certManager.repository`](https://github.com/deployKF/deployKF/blob/v0.1.0/generator/default_values.yaml#L79)_
2. This resulted in a significant load on the ArgoCD repo-server, which was constantly pulling the upstream repos every few seconds
3. This introduced an unacceptable risk of users being unable to "sync" their manifests during a GitHub outage

Finally, to address the concerns about flexibility, from the end-user perspective, the generated manifests are meant to be a black box. That is, people shouldn't really be changing them after they run the generator. The goal is to have a value for anything which the user may want to change (so users don't have to touch the resource YAML).

From the project's perspective, as you note, there is a [`sync_upstream.sh` script](https://github.com/deployKF/deployKF/blob/v0.1.0/generator/templates/manifests/kubeflow-tools/pipelines/sync_upstream.sh) in each vendored folder, and the `UPSTREAM_REF` can easily be changed to newer ones if desired, and this even allows us to see in our git-diff what changed.

**Reply by jaklan** [reply]
@thesuperzapper thanks for the answer. Correct me if I'm wrong, but with that approach - I would need to fork the repo, modify the `UPSTREAM_REF` value in `sync_upstream.sh` and run the script to pull the version of manifests I want to use (or just modify the `upstream` directories in any other way)?

**Reply by thesuperzapper** [reply]
@jaklan I am a bit confused about what you are asking.

The intention of deployKF is that users treat the generator package as a black box, and only modify the values they are passing to the generator, which is a versioned zip package for each [release of deployKF](https://github.com/deployKF/deployKF/releases/tag/v0.1.0).

There should never be a situation where an end-user needs to modify this repo (unless they are contributing). Even if a user wanted to modify `UPSTREAM_REF`, they would very likely break the generator package, because there is no guarantee our kustomize patches (which are configured by values) would still work as expected with different upstream manifests.

Those `sync_upstream.sh` scripts are for contributors of deployKF, they make it easier to update deployKF with support for newer versions of the upstream tools, as they are released.

**Reply by jaklan** [reply]
@thesuperzapper I'm asking how I could select version or simply modify the upstream manifests when using `deployKF`. But if users are expected to treat them as a "black box" and don't interact with them anyhow, I believe that answers my question  - I just can't.

**Reply by thesuperzapper** [reply]
@jaklan for reference, was there something specific you could not achieve using the available values?

Or was this a pure hypothetical?

**Reply by jaklan** [reply]
@thesuperzapper for example we have plenty of own patches in `kubeflow/pipelines/kustomization.yaml` (`patchesStrategicMerge`). How could they be applied when using `deployKF`?

**Reply by thesuperzapper** [reply]
Can you clarify what your custom patches do? This will help me point you to the equivalent deployKF values (or help me find new values which need to be created, which will help the project).

**Reply by jaklan** [reply]
@kromanow94 could you jump in? You would be the best person to describe that in details

**Reply by kromanow94** [reply]
Hi @jaklan , @thesuperzapper .

We have modifications for:
* oauth2-proxy (oidc-authservice is fully removed as well as the EnvoyFilters for auth),
* istio tweaks,
* AWS S3/MinioGW (because we don't want to use iam user credentials, this could be probably improved now as I saw AWS Labs KF Helm Charts don't have the iam user credentials anymore),
* AWS RDS,
* disabling kubeflow cache on a platform level,
* modifications related to moving ConfigMaps and Secrets from the upstream kustomize overlays to AWS CDK for parametrization.

**Comment by thesuperzapper** [comment]
@kromanow94 All of these are currently possible with deployKF values.

> * oauth2-proxy (oidc-authservice is fully removed as well as the EnvoyFilters for auth)

This is already how deployKF auth works, if you want a taste of how the oauth2-proxy and istio have been integrated, see the [`EnvoyFilters` under the `deploykf-istio-gateway`](https://github.com/deployKF/deployKF/tree/main/generator/templates/manifests/deploykf-core/deploykf-istio-gateway/templates/gateway) app.

> * istio tweaks

Not clear what is being referred to here.

> * AWS S3/MinioGW (because we don't want to use iam user credentials, this could be probably improved now as I saw AWS Labs KF Helm Charts don't have the iam user credentials anymore)

deployKF does not use the minio s3-gateway (because it is no longer supported by MinIO, and poses security risks), but you can easily connect Kubeflow Pipelines to S3 using the following values:

- [`kubeflow_tools.pipelines.objectStore`](https://github.com/deployKF/deployKF/blob/v0.1.0/generator/default_values.yaml#L1640-L1649) to set your S3 host URL and disable minio
- [`kubeflow_tools.pipelines.objectStore.auth`](https://github.com/deployKF/deployKF/blob/v0.1.0/generator/default_values.yaml#L1651-L1675) to set KFP backend auth
- [`deploykf_core.deploykf_profiles_generator.profileDefaults.tools.kubeflowPipelines.objectStoreAuth`](https://github.com/deployKF/deployKF/blob/v0.1.0/generator/default_values.yaml#L717-L745)to set KFP profile auth

> * AWS RDS

You can connect Kubeflow Pipelines to RDS using the [`kubeflow_tools.pipelines.mysql`](https://github.com/deployKF/deployKF/blob/v0.1.0/generator/default_values.yaml#L1677-L1691) values.

> * disabling kubeflow cache on a platform level

You can disable the KFP (v1) cache using the following values under [`kubeflow_tools.pipelines.cache`](https://github.com/deployKF/deployKF/blob/v0.1.0/generator/default_values.yaml#L1745-L1753):

- `kubeflow_tools.pipelines.cache.maximumMaxCacheStaleness`
- `kubeflow_tools.pipelines.cache.defaultMaxCacheStaleness`

> * modifications related to moving ConfigMaps and Secrets from the upstream kustomize overlays to AWS CDK for parametrization.

Since everything is now configured with values, you won't need to think about this anymore.

**Reply by kromanow94** [reply]
Great, thanks for explaining.
